modules/programs/afew.nix

```
@@ -6,21 +6,134 @@ let
 
   cfg = config.programs.afew;
 
+  dag = config.lib.dag;
+
+  mkIniKeyValue = key: value:
+    let
+      tweakVal = v:
+        if isString v then v
+        else if isList v then concatMapStringsSep ";" tweakVal v
+        else if isBool v then (if v then "true" else "false")
+        else toString v;
+    in
+      "${key}=${tweakVal value}";
+
+  mkDag = filters:
+    let
+      mkNode = nodeName: filter:
+        let
+          before = (filter.before or cfg.defaultBefore);
+          after = (filter.after or cfg.defaultAfter);
+          filterName = (filter.class or "Filter");
+          filterKV = (removeAttrs filter ["after" "before" "class"]) // {
+            "#nodeName" = nodeName;
+            "#before" = before;
+            "#after" = after;
+          };
+        in
+          dag.entryBetween before after (nameValuePair filterName filterKV);
+    in
+      mapAttrs mkNode filters;
+
+  /* Return a topologically sorted list of DAG entries encapsulating
+     name value pairs.
+  */
+  reorderFilters = filters:
+    (dag.topoSort (mkDag filters)).result;
+
+  /* Convert a topologically sorted list of DAG nodes to a list of INI
+     sections.  The section name is appended with an index number to
+     prevent clashes, as is required by afew.
+  */
+  toOrderedINI = {
+    mkKeyValue ? mkKeyValueDefault {} "="
+  }: filters:
+    let
+      mkSection = iota: node: ''
+        [${node.data.name}.${toString iota}]
+        ${toKV node.data.value}
+      '';
+      toKV = attrs: generators.toKeyValue { inherit mkKeyValue; } attrs;
+    in
+      concatStringsSep "\n" (imap0 mkSection filters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    defaultBefore = mkOption {
+      type = types.listOf types.string;
+      default = ["@end"];
+      description = ''
+        List of node names in the DAG that filters should appear
+        before by default.
+      '';
+    };
+
+    defaultAfter = mkOption {
+      type = types.listOf types.string;
+      default = ["@begin"];
+      description = ''
+        List of node names in the DAG that filters should appear after
+        by default.
+      '';
+    };
+
+    defaultFilters = mkOption {
+      type = types.attrsOf types.attrs;
+      default = {
+        "@begin" = { message = "BEGIN"; after = []; before = []; };
+        "@end" = { message = "END"; after = ["@begin"]; before = []; };
+      };
+      description = ''
+        An initial value for the filters DAG. It should only contain
+        "dummy" nodes. Those dummy nodes are conventionally prefixed
+        with @.
+      '';
+    };
+
+    filters = mkOption {
+      type = types.attrsOf types.attrs;
```

I think the type here could be changed as per
```suggestion
      type = with types;
        let
          primitive = either bool (either int str);
        in
          attrsOf (attrsOf (either primitive (listOf str)));
```
Perhaps `primitive` could even be `primitive = either bool str`?

---

As afew can be user-extendable, I'm not sure what the allowed types of the leaves should be.
I think we could go with `(either primitive (listOf primitive))`.